Test case where some inputs are Tensor Subclasses in CompositeCompiance

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #74646
* __->__ #74645
* #74644

This PR adds tests for when only some inputs are Tensor Subclasses.

Why is this important to test?
==============================

Consider the following hypothetical out-of-place operation:
```
def my_add(x, y):
  result = x.clone()
  result.add_(y)
  return result
```

You may expect this to work the same as torch.add. If x is not a Tensor
Subclass, but y is a Tensor subclass, then this returns us a regular
Tensor, NOT a Tensor subclass!

This is exactly the type of in-place operations that causes `vmap` to
fail and will be problematic for certain Tensor Subclasses in the future
so we're adding tests to make sure Composite pytorch operations don't do
this.

What exactly does this PR do?
=============================
Composite compliance now takes a sample input and produces a test case
where some of the sample inputs are Tensor Subclasses. It then sends
this through the original operation, once with Python Mode and one
without.

(Why once with Python Mode? Because we want to use it to detect the
pattern of "create a Tensor and call resize_ on it")

Finally, it repeats this process for all possiblities where the inputs
are Tensor subclasses. For example, if the sample input is (x, y), then
we test all four of the following cases:
- Subclass(x), y
- x, Subclass(y)
- Subclass(x), Subclass(y)
- x, y

Test Plan
=========
- run tests

Differential Revision: [D35186862](https://our.internmc.facebook.com/intern/diff/D35186862)

zou3519 committed:
Test case where some inputs are Tensor Subclasses in CompositeCompiance

This PR adds tests for when only some inputs are Tensor Subclasses.

Why is this important to test?
==============================

Consider the following hypothetical out-of-place operation:
```
def my_add(x, y):
  result = x.clone()
  result.add_(y)
  return result
```

You may expect this to work the same as torch.add. If x is not a Tensor
Subclass, but y is a Tensor subclass, then this returns us a regular
Tensor, NOT a Tensor subclass!

This is exactly the type of in-place operations that causes `vmap` to
fail and will be problematic for certain Tensor Subclasses in the future
so we're adding tests to make sure Composite pytorch operations don't do
this.

What exactly does this PR do?
=============================
Composite compliance now takes a sample input and produces a test case
where some of the sample inputs are Tensor Subclasses. It then sends
this through the original operation, once with Python Mode and one
without.

(Why once with Python Mode? Because we want to use it to detect the
pattern of "create a Tensor and call resize_ on it")

Finally, it repeats this process for all possiblities where the inputs
are Tensor subclasses. For example, if the sample input is (x, y), then
we test all four of the following cases:
- Subclass(x), y
- x, Subclass(y)
- Subclass(x), Subclass(y)
- x, y

Test Plan
=========
- run tests

[ghstack-poisoned]

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -173,6 +192,107 @@ def wrap(e):
+        return False
+    if len(lst) == 0:
+        return False
+    return isinstance(lst[0], torch.Tensor)
```
Why not `all(isinstance(t, torch.Tensor) for t in lst)`

zou3519 replied:
Sure

albanD commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -173,6 +192,107 @@ def wrap(e):
+# [CCT(A), 1, B]
+# [A, 1, CCT(B)]
+# [A, 1, B]
+# NB: Yes, this is exponential. No, we don't care too much because PyTorch ops
```
Are these slow enough that we should consider reducing the number of devices we run on or run them less often?

zou3519 replied:
They're not that slow (these tests run in under a minute total on my machine) and people will exile individual tests to the slow tests section if they notice they are slow

zou3519 replied:
I ran the stats and recorded the actual information (the one minute ballpark I cited before was from memory)

Total runtime is 1:14 on both CPU and CUDA.
CPU-only is: 32s

zou3519 commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -173,6 +192,107 @@ def wrap(e):
+
+        try:
+            op(*new_args, **new_kwargs)
+        except RuntimeError as err:
```
> any error that might happen when a function is not properly implemented?

We don't really want to catch errors where a function has a bug in it, but we do because of the following.

There's two things we want to catch:
- errors that would raise within the torch_dispatch impl
- data_ptr accesses

The first is easy to filter for (we could make the error a different error class), the second is always going to be a RuntimeError due to how it is implemented (if you try to access the data_ptr of the wrapper Tensor, it raises you some internal RuntimeError).

So the most general thing to catch here was RuntimeError

albanD replied:
Ok!
A small comment here specifying that we want both of these captured would be nice.

zou3519 commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -92,6 +93,12 @@ def is_inplace_view_fn(func):
+
+# Introspection please save us
+def is_inplace(func):
+    return func.overloadpacket.__name__[-1] == '_'
```
Yes, I will add those cases. I don't know if this catches everything but we're trying to be best effort here until there is a better solution

albanD replied:
Yes, the tags should provide a better solution here when they're ready cc @anjali411

ezyang commented on torch/testing/_internal/composite_compliance.py:
```diff
@@ -187,11 +307,4 @@ def wrap(e):
-                           "through the CompositeImplicitAutograd Compliance section in "
-                           "aten/src/ATen/native/README.md for how to resolve this. "
-                           ) from err
+        raise_composite_compliance_error(err)
```
How long do the new tests take to run?

zou3519 replied:
Total runtime is 1:14 on both CPU and CUDA.
CPU-only is: 32s

ezyang reviewed (approved):
This makes me wonder if testing is the wrong way to check these properties. If these tests are quick LGTM.

zou3519 commented:
> If these tests are quick LGTM.

@ezyang this is one of the reasons why I wanted symbolic shape stuff for OpInfos -- some of these OpInfos are too large and we really don't need them to be so large for these tests :).

> This makes me wonder if testing is the wrong way to check these properties. 

I don't think testing as done in this PR is the optimal way to check these properties due to the exponential runtime. I don't have any better ideas for how to test it with our current infrastructure (and this is exactly what we do for vmap testing -- we have to test the cases where some inputs are BatchedTensors and some aren't...). Here are the things I would want in an ideal world:
- If we, for example, had a Python source of truth for our composite operations then we could statically analyze the Python and that should be faster
- Alternatively, if we enforced that out-of-place operations be written in functional form but then used a backend or a "mutation pass" (the opposite of a functionalization pass) to generate fast kernels that do end up doing in-place operations, then we wouldn't even need to test this, because what we would be transforming over are the out-of-place operations

albanD commented:
> If we, for example, had a Python source of truth for our composite operations then we could statically analyze the Python and that should be faster


I'm not sure about this one though. The same test would be needed to ensure that the source of truth does match the c++ impl (unless you generate the c++ from the python of course).

zou3519 committed:
Update on "Test case where some inputs are Tensor Subclasses in CompositeCompiance"


This PR adds tests for when only some inputs are Tensor Subclasses.

Why is this important to test?
==============================

Consider the following hypothetical out-of-place operation:
```
def my_add(x, y):
  result = x.clone()
  result.add_(y)
  return result
```

You may expect this to work the same as torch.add. If x is not a Tensor
Subclass, but y is a Tensor subclass, then this returns us a regular
Tensor, NOT a Tensor subclass!

This is exactly the type of in-place operations that causes `vmap` to
fail and will be problematic for certain Tensor Subclasses in the future
so we're adding tests to make sure Composite pytorch operations don't do
this.

What exactly does this PR do?
=============================
Composite compliance now takes a sample input and produces a test case
where some of the sample inputs are Tensor Subclasses. It then sends
this through the original operation, once with Python Mode and one
without.

(Why once with Python Mode? Because we want to use it to detect the
pattern of "create a Tensor and call resize_ on it")

Finally, it repeats this process for all possiblities where the inputs
are Tensor subclasses. For example, if the sample input is (x, y), then
we test all four of the following cases:
- Subclass(x), y
- x, Subclass(y)
- Subclass(x), Subclass(y)
- x, y

Test Plan
=========
- run tests

[ghstack-poisoned]

zou3519 committed:
Update on "Test case where some inputs are Tensor Subclasses in CompositeCompiance"


This PR adds tests for when only some inputs are Tensor Subclasses.

Why is this important to test?
==============================

Consider the following hypothetical out-of-place operation:
```
def my_add(x, y):
  result = x.clone()
  result.add_(y)
  return result
```

You may expect this to work the same as torch.add. If x is not a Tensor
Subclass, but y is a Tensor subclass, then this returns us a regular
Tensor, NOT a Tensor subclass!

This is exactly the type of in-place operations that causes `vmap` to
fail and will be problematic for certain Tensor Subclasses in the future
so we're adding tests to make sure Composite pytorch operations don't do
this.

What exactly does this PR do?
=============================
Composite compliance now takes a sample input and produces a test case
where some of the sample inputs are Tensor Subclasses. It then sends
this through the original operation, once with Python Mode and one
without.

(Why once with Python Mode? Because we want to use it to detect the
pattern of "create a Tensor and call resize_ on it")

Finally, it repeats this process for all possiblities where the inputs
are Tensor subclasses. For example, if the sample input is (x, y), then
we test all four of the following cases:
- Subclass(x), y
- x, Subclass(y)
- Subclass(x), Subclass(y)
- x, y

Test Plan
=========
- run tests

[ghstack-poisoned]

zou3519 commented:
@zou3519 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35186862).
